File: dsa_leetcode_beginner/0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.cpp
#include <algorithm>
#include <climits>

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    int maxPathSum(TreeNode* root) {
        int maxSum = INT_MIN;
        dfs(root, maxSum);
        return maxSum;
    }

private:
    int dfs(TreeNode* node, int& maxSum) {
        if (node == nullptr) {
            return 0;
        }

        int leftGain = std::max(dfs(node->left, maxSum), 0);
        int rightGain = std::max(dfs(node->right, maxSum), 0);

        int priceNewPath = node->val + leftGain + rightGain;
        maxSum = std::max(maxSum, priceNewPath);

        return node->val + std::max(leftGain, rightGain);
    }
};

File: dsa_leetcode_beginner/0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        max_sum = float('-inf')

        def dfs(node):
            if not node:
                return 0

            # Recursive call on left and right child
            left_gain = max(dfs(node.left), 0)
            right_gain = max(dfs(node.right), 0)

            # Update max_sum if it's better to start a new path
            price_newpath = node.val + left_gain + right_gain
            max_sum = max(max_sum, price_newpath)

            # For recursion return the max gain if continue the same path
            return node.val + max(left_gain, right_gain)

        dfs(root)
        return max_sum
        
        
#Question: https://leetcode.com/problems/binary-tree-maximum-path-sum
#Blog: https://blog.unwiredlearning.com/binary-tree-maximum-path-sum